src/cargo/ops/cargo_compile.rs

```diff
@@ -857,7 +864,21 @@ fn generate_targets(
             let features_for = FeaturesFor::from_for_host(target.proc_macro());
             let features = resolved_features.activated_features(pkg.package_id(), features_for);
 
-            for kind in requested_kinds {
+            // If `--target` has not been specified, then the unit
+            // graph is built almost like if `--target $HOST` was
+            // specified. See `rebuild_unit_graph_shared` for more on
+            // why this is done. However, if the package has its own
+            // `package.target` key, then this gets used instead of
+            // `$HOST`
+            let explicit_kinds: Vec<_> = requested_kinds
+                .iter()
+                .map(|kind| match kind {
+                    CompileKind::Host => pkg.manifest().kind().unwrap_or(explicit_host_kind),
```

Comment:
I'm not 100% sure on this interaction myself. For example if you have an application that embeds a wasm blob, then you probably want the wasm blob always compiled as wasm but the application itself may want to get compiled for different architectures. I could also see, though, use cases where `target` is the default but it should still be overridable with `--target`.

Basically I don't think there's a great answer at this time and we'll likely want to tweak the design here.

Reply:
Yes, this is also one thing I was wondering about. An idea I had was to add `default-target` and `forced-target` instead of adding `target`, and then it'd look like this:
```
pkg.manifest().forced_kind().unwrap_or(match kind {
    CompileKind::Host => pkg.manifest().default_kind().unwrap_or(explicit_host_kind),
    // ...
})
```

How would that sound? To me it feels like it's a bit more complex to use, so I was wondering which use case should be prioritized, and so for the initial draft I went with “just reuse the behavior from `.cargo/config` of making it a default” ; but it wasn't because I have found a good reason either way — actually now that I think about it more I kind of believe that if only one of default/forced should be made available, it might be better for it to be forced, as I can't think of good use cases where `target` should be a default while also being per-crate instead of in `.cargo/config`

Reply:
Indeed yeah having both keys in the manifest would solve this issue. I'm not sure if it's the best design but this is all unstable anyway so we can tweak it later.